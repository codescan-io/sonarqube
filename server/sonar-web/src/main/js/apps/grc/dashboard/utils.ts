
import { RawHotspot, RiskExposure } from "../../../types/security-hotspots";
import { groupBy } from 'lodash';
import { groupByCategory, RISK_EXPOSURE_LEVELS } from "../../security-hotspots/utils";

export function getReadableDateFormat(date:string){
    let t:Date = new Date(date);
    return t.toLocaleString('en-US', { year:'numeric', month:'numeric', day:'numeric',hour: 'numeric', minute: 'numeric', hour12: true });
}

export function getQualityGateInfo(event:T.Analysis) :string|undefined
{
    const qualityGateEvent = event.events.find(event => event.category === 'QUALITY_GATE');
    return qualityGateEvent?.name;
}

export function getQualityProfileInfo(event:T.Analysis) :string|undefined
{
    const qualityProfileEvent = event.events.find(event => event.category === 'QUALITY_PROFILE');
    return qualityProfileEvent?.name;
}

export function getVersionInfo(event:T.Analysis) :string|undefined
{
    const versionEvent = event.events.find(event => event.category === 'VERSION');
    return versionEvent?.name;
}

export function getHotspotsBasedOnRiskExposure(hotspots:RawHotspot[]):{high:number,medium:number,low:number}{
    return {
        "high": hotspots.filter((hotspot:RawHotspot)=>hotspot.vulnerabilityProbability === RiskExposure.HIGH).length,
        "medium": hotspots.filter((hotspot:RawHotspot)=>hotspot.vulnerabilityProbability === RiskExposure.MEDIUM).length,
        "low": hotspots.filter((hotspot:RawHotspot)=>hotspot.vulnerabilityProbability === RiskExposure.LOW).length
    };
}

export function getGroupHotspots(hotspots: RawHotspot[], securityCategories: T.StandardSecurityCategories){
    const risks = groupBy(hotspots, h => h.vulnerabilityProbability);

    return RISK_EXPOSURE_LEVELS.map(risk => ({
        risk,
        categories: groupByCategory(risks[risk], securityCategories)
      })).filter(risk => risk.categories.length > 0);
}
