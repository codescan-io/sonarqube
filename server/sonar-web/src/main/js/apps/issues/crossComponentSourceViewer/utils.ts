/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { ComponentQualifier } from '../../../types/component';
import {
  ExpandDirection,
  FlowLocation,
  Issue,
  LineMap,
  Snippet,
  SnippetGroup,
  SnippetsByComponent,
  SourceLine,
} from '../../../types/types';

const LINES_ABOVE = 5;
const LINES_BELOW = 5;
export const MERGE_DISTANCE = 11; // Merge if snippets are eleven lines away (separated by 10 lines) or fewer
export const LINES_BELOW_ISSUE = 9;
export const EXPAND_BY_LINES = 50;

function unknownComponent(key: string): SnippetsByComponent {
  return {
    component: {
      key,
      measures: {},
      path: '',
      project: '',
      projectName: '',
      q: ComponentQualifier.File,
      uuid: '',
    },
    sources: [],
  };
}

function collision([startA, endA]: number[], [startB, endB]: number[]) {
  return !(startA > endB + MERGE_DISTANCE || endA < startB - MERGE_DISTANCE);
}

export function getPrimaryLocation(issue: Issue): FlowLocation {
  return {
    component: issue.component,
    textRange: issue.textRange || {
      endLine: 0,
      endOffset: 0,
      startLine: 0,
      startOffset: 0,
    },
  };
}

function addLinesBellow(params: { issue: Issue; locationEnd: number }) {
  const { issue, locationEnd } = params;
  const issueEndLine = (issue.textRange && issue.textRange.endLine) || 0;

  if (!issueEndLine || issueEndLine === locationEnd) {
    return locationEnd + LINES_BELOW_ISSUE;
  }

  return locationEnd + LINES_BELOW;
}

export function createSnippets(params: {
  component: string;
  issue: Issue;
  locations: FlowLocation[];
}): Snippet[] {
  const { component, issue, locations } = params;

  const hasSecondaryLocations = issue.secondaryLocations.length > 0;
  const addIssueLocation =
    hasSecondaryLocations && issue.component === component && issue.textRange !== undefined;

  // For each location: compute its range, and then compare with other ranges
  // to merge snippets that collide.
  const ranges = (addIssueLocation ? [getPrimaryLocation(issue), ...locations] : locations).reduce(
    (snippets: Snippet[], loc, index) => {
      const startIndex = Math.max(1, loc.textRange.startLine - LINES_ABOVE);
      const endIndex = addLinesBellow({ issue, locationEnd: loc.textRange.endLine });

      let firstCollision: { start: number; end: number } | undefined;

      // Remove ranges that collide into the first collision
      snippets = snippets.filter((snippet) => {
        if (collision([snippet.start, snippet.end], [startIndex, endIndex])) {
          let keep = false;
          // Check if we've already collided
          if (!firstCollision) {
            firstCollision = snippet;
            keep = true;
          }
          // Merge with first collision:
          firstCollision.start = Math.min(startIndex, snippet.start, firstCollision.start);
          firstCollision.end = Math.max(endIndex, snippet.end, firstCollision.end);

          // remove the range if it was not the first collision
          return keep;
        }
        return true;
      });

      if (firstCollision === undefined) {
        snippets.push({
          start: startIndex,
          end: endIndex,
          index,
        });
      }

      return snippets;
    },
    []
  );

  // Sort snippets by line number in the case of secondary locations
  // Preserve location order for flows!
  return hasSecondaryLocations ? ranges.sort((a, b) => a.start - b.start) : ranges;
}

export function linesForSnippets(snippets: Snippet[], componentLines: LineMap) {
  return snippets
    .map((snippet) => {
      const lines = [];
      for (let i = snippet.start; i <= snippet.end; i++) {
        if (componentLines[i]) {
          lines.push(componentLines[i]);
        }
      }
      return lines;
    })
    .filter((snippet) => snippet.length > 0);
}

export function groupLocationsByComponent(
  issue: Issue,
  locations: FlowLocation[],
  components: { [key: string]: SnippetsByComponent }
) {
  let currentComponent = '';
  let currentGroup: SnippetGroup;
  const groups: SnippetGroup[] = [];

  const addGroup = (componentKey: string) => {
    currentGroup = {
      ...(components[componentKey] || unknownComponent(componentKey)),
      locations: [],
    };
    groups.push(currentGroup);
    currentComponent = componentKey;
  };

  if (
    issue.secondaryLocations.length > 0 &&
    locations.every((loc) => loc.component !== issue.component)
  ) {
    addGroup(issue.component);
  }

  locations.forEach((loc, index) => {
    if (loc.component !== currentComponent) {
      addGroup(loc.component);
    }
    loc.index = index;
    currentGroup.locations.push(loc);
  });

  if (groups.length === 0) {
    groups.push({ locations: [], ...components[issue.component] });
  }

  return groups;
}

export function expandSnippet({
  direction,
  snippetIndex,
  snippets,
}: {
  direction: ExpandDirection;
  snippetIndex: number;
  snippets: Snippet[];
}) {
  const snippetToExpand = snippets.find((s) => s.index === snippetIndex);
  if (!snippetToExpand) {
    throw new Error(`Snippet ${snippetIndex} not found`);
  }

  snippetToExpand.start = Math.max(
    0,
    snippetToExpand.start - (direction === 'up' ? EXPAND_BY_LINES : 0)
  );
  snippetToExpand.end += direction === 'down' ? EXPAND_BY_LINES : 0;

  return snippets.map((snippet) => {
    if (snippet.index === snippetIndex) {
      return snippetToExpand;
    }
    if (collision([snippet.start, snippet.end], [snippetToExpand.start, snippetToExpand.end])) {
      // Merge with expanded snippet
      snippetToExpand.start = Math.min(snippet.start, snippetToExpand.start);
      snippetToExpand.end = Math.max(snippet.end, snippetToExpand.end);
      snippet.toDelete = true;
    }
    return snippet;
  });
}

export function inSnippet(line: number, snippet: SourceLine[]) {
  return line >= snippet[0].line && line <= snippet[snippet.length - 1].line;
}
