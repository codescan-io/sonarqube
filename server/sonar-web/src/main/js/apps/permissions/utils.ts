/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { translate } from '../../helpers/l10n';
import { Dict, PermissionDefinition, PermissionDefinitionGroup } from '../../types/types';

export const PERMISSIONS_ORDER_FOR_PROJECT_TEMPLATE = [
  'user',
  'codeviewer',
  'issueadmin',
  'securityhotspotadmin',
  'admin',
  'scan',
];

export const PERMISSIONS_ORDER_GLOBAL = [
  'admin',
  { category: 'administer', permissions: ['gateadmin', 'profileadmin'] },
  'scan',
  { category: 'creator', permissions: ['provisioning', 'applicationcreator', 'portfoliocreator'] },
];

export const PERMISSIONS_ORDER_FOR_VIEW = ['user', 'admin'];

export const PERMISSIONS_ORDER_FOR_DEV = ['user', 'admin'];

export const PERMISSIONS_ORDER_BY_QUALIFIER: Dict<string[]> = {
  TRK: PERMISSIONS_ORDER_FOR_PROJECT_TEMPLATE,
  VW: PERMISSIONS_ORDER_FOR_VIEW,
  SVW: PERMISSIONS_ORDER_FOR_VIEW,
  APP: PERMISSIONS_ORDER_FOR_VIEW,
  DEV: PERMISSIONS_ORDER_FOR_DEV,
};

function convertToPermissionDefinition(permission: string, l10nPrefix: string) {
  const name = translate(`${l10nPrefix}.${permission}`);
  const description = translate(`${l10nPrefix}.${permission}.desc`);

  return {
    key: permission,
    name,
    description,
  };
}

export function filterPermissions(
  permissions: Array<string | { category: string; permissions: string[] }>,
  hasApplicationsEnabled: boolean,
  hasPortfoliosEnabled: boolean
) {
  return permissions.map((permission) => {
    if (typeof permission === 'object' && permission.category === 'creator') {
      return {
        ...permission,
        permissions: permission.permissions.filter((p) => {
          return (
            p === 'provisioning' ||
            (p === 'portfoliocreator' && hasPortfoliosEnabled) ||
            (p === 'applicationcreator' && hasApplicationsEnabled)
          );
        }),
      };
    }
    return permission;
  });
}

export function convertToPermissionDefinitions(
  permissions: Array<string | { category: string; permissions: string[] }>,
  l10nPrefix: string
): Array<PermissionDefinition | PermissionDefinitionGroup> {
  return permissions.map((permission) => {
    if (typeof permission === 'object') {
      return {
        category: permission.category,
        permissions: permission.permissions.map((permission) =>
          convertToPermissionDefinition(permission, l10nPrefix)
        ),
      };
    }
    return convertToPermissionDefinition(permission, l10nPrefix);
  });
}

export function isPermissionDefinitionGroup(
  permission?: PermissionDefinition | PermissionDefinitionGroup
): permission is PermissionDefinitionGroup {
  return Boolean(permission && (permission as PermissionDefinitionGroup).category);
}
