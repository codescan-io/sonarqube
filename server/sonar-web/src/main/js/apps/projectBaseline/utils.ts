/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { isNewCodeDefinitionCompliant } from '../../helpers/periods';
import { NewCodePeriodSettingType } from '../../types/types';

export function getSettingValue({
  analysis,
  days,
  referenceBranch,
  type,
}: {
  analysis?: string;
  days?: string;
  referenceBranch?: string;
  type?: NewCodePeriodSettingType;
}) {
  switch (type) {
    case NewCodePeriodSettingType.NUMBER_OF_DAYS:
      return days;
    case NewCodePeriodSettingType.REFERENCE_BRANCH:
      return referenceBranch;
    case NewCodePeriodSettingType.SPECIFIC_ANALYSIS:
      return analysis;
    default:
      return undefined;
  }
}

export function validateSetting(state: {
  analysis?: string;
  currentSetting?: NewCodePeriodSettingType;
  currentSettingValue?: string;
  days: string;
  overrideGeneralSetting?: boolean;
  referenceBranch?: string;
  selected?: NewCodePeriodSettingType;
}) {
  const {
    analysis = '',
    currentSetting,
    currentSettingValue,
    days,
    overrideGeneralSetting,
    referenceBranch = '',
    selected,
  } = state;

  let isChanged;
  if (!currentSetting && overrideGeneralSetting !== undefined) {
    isChanged = overrideGeneralSetting;
  } else {
    isChanged =
      overrideGeneralSetting === false ||
      selected !== currentSetting ||
      (selected === NewCodePeriodSettingType.NUMBER_OF_DAYS && days !== currentSettingValue) ||
      (selected === NewCodePeriodSettingType.SPECIFIC_ANALYSIS &&
        analysis !== currentSettingValue) ||
      (selected === NewCodePeriodSettingType.REFERENCE_BRANCH &&
        referenceBranch !== currentSettingValue);
  }

  const isValid =
    overrideGeneralSetting === false ||
    (!!selected &&
      isNewCodeDefinitionCompliant({
        type: selected,
        value: days,
      }) &&
      (selected !== NewCodePeriodSettingType.SPECIFIC_ANALYSIS || analysis.length > 0) &&
      (selected !== NewCodePeriodSettingType.REFERENCE_BRANCH || referenceBranch.length > 0));

  return { isChanged, isValid };
}
