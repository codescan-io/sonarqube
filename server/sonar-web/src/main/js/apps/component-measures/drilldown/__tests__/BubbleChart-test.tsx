/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import userEvent from '@testing-library/user-event';
import { MetricsEnum } from 'design-system';
import { keyBy } from 'lodash';
import React from 'react';
import { AutoSizerProps } from 'react-virtualized';
import { byRole, byText } from 'testing-library-selector';
import { mockComponentMeasure } from '../../../../helpers/mocks/component';
import { mockMeasure, mockMetric, mockPaging } from '../../../../helpers/testMocks';
import { renderComponent } from '../../../../helpers/testReactTestingUtils';
import { MetricKey, MetricType } from '../../../../types/metrics';
import { enhanceComponent } from '../../utils';
import BubbleChart from '../BubbleChart';

jest.mock('react-virtualized/dist/commonjs/AutoSizer', () => ({
  AutoSizer: ({ children }: AutoSizerProps) => children({ width: 100, height: NaN }),
}));

jest.mock('d3-zoom', () => ({
  zoom: jest.fn().mockReturnValue({ scaleExtent: jest.fn().mockReturnValue({ on: jest.fn() }) }),
}));

jest.mock('d3-selection', () => ({
  select: jest.fn().mockReturnValue({ call: jest.fn() }),
}));

const metrics = keyBy(
  [
    mockMetric({ key: MetricKey.ncloc, type: MetricType.Integer }),
    mockMetric({ key: MetricKey.security_remediation_effort, type: MetricType.Data }),
    mockMetric({ key: MetricKey.vulnerabilities, type: MetricType.Integer }),
    mockMetric({ key: MetricKey.security_rating, type: MetricType.Rating }),
  ],
  (m) => m.key
);

const ui = {
  show500Files: byText(/component_measures.legend.only_first_500_files/),
  bubbles: byRole('link', { name: '' }),
  filterCheckbox: (name: string) => byRole('checkbox', { name }),
};

it('should render correctly', async () => {
  renderBubbleChart();

  expect(await ui.show500Files.find()).toBeInTheDocument();
  expect(ui.bubbles.getAll()).toHaveLength(1);
});

it('should filter by rating', async () => {
  const user = userEvent.setup();
  renderBubbleChart();

  expect(await ui.bubbles.findAll()).toHaveLength(1);
  Object.keys(MetricsEnum).forEach((rating) => {
    expect(ui.filterCheckbox(rating).get()).toBeInTheDocument();
  });

  await user.click(ui.filterCheckbox(MetricsEnum.C).get());
  expect(ui.bubbles.getAll()).toHaveLength(1);
});

function renderBubbleChart(overrides: Partial<BubbleChart['props']> = {}) {
  return renderComponent(
    <BubbleChart
      componentKey="foo"
      components={[
        enhanceComponent(
          mockComponentMeasure(true, {
            measures: [
              mockMeasure({ value: '236', metric: MetricKey.ncloc }),
              mockMeasure({ value: '10', metric: MetricKey.security_remediation_effort }),
              mockMeasure({ value: '3', metric: MetricKey.vulnerabilities }),
              mockMeasure({ value: '2', metric: MetricKey.security_rating }),
            ],
          }),
          metrics[MetricKey.vulnerabilities],
          metrics
        ),
      ]}
      domain="Security"
      metrics={metrics}
      paging={mockPaging({ total: 1000 })}
      updateSelected={jest.fn()}
      {...overrides}
    />
  );
}
