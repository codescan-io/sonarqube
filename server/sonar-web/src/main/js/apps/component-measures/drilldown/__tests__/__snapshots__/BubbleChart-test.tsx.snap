// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly: all on x=0 1`] = `
<div
  className="measure-overview-bubble-chart"
>
  <div
    className="measure-overview-bubble-chart-header"
  >
    <span
      className="measure-overview-bubble-chart-title"
    >
      <div
        className="display-flex-center"
      >
        component_measures.domain_x_overview.Security
        <HelpTooltip
          className="spacer-left"
          overlay={null}
        />
      </div>
    </span>
    <span
      className="measure-overview-bubble-chart-legend"
    >
      <span
        className="note"
      >
        <span
          className="spacer-right"
        >
          component_measures.legend.color_x.security_rating
        </span>
        component_measures.legend.size_x.vulnerabilities
      </span>
      <ColorRatingsLegend
        className="spacer-top"
        filters={{}}
        onRatingClick={[Function]}
      />
    </span>
  </div>
  <div
    className="measure-overview-bubble-chart-content"
  >
    <div
      className="text-center small spacer-top spacer-bottom"
    >
      <ForwardRef(Link)
        to={
          {
            "pathname": "/component_measures",
            "search": "?id=foo&metric=vulnerabilities&view=list",
          }
        }
      >
        component_measures.overview.see_data_as_list
      </ForwardRef(Link)>
    </div>
    <BubbleChart
      displayXGrid={true}
      displayXTicks={true}
      displayYGrid={true}
      displayYTicks={true}
      formatXTick={[Function]}
      formatYTick={[Function]}
      height={500}
      items={
        [
          {
            "color": undefined,
            "data": {
              "key": "foo:src/index.tsx",
              "leak": "1.0",
              "measures": [
                {
                  "bestValue": true,
                  "leak": "1.0",
                  "metric": {
                    "id": "ncloc",
                    "key": "ncloc",
                    "name": "ncloc",
                    "type": "INT",
                  },
                  "period": {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                  "value": "0",
                },
                {
                  "bestValue": true,
                  "leak": "1.0",
                  "metric": {
                    "id": "security_remediation_effort",
                    "key": "security_remediation_effort",
                    "name": "security_remediation_effort",
                    "type": "DATA",
                  },
                  "period": {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                  "value": "0",
                },
                {
                  "bestValue": true,
                  "leak": "1.0",
                  "metric": {
                    "id": "vulnerabilities",
                    "key": "vulnerabilities",
                    "name": "vulnerabilities",
                    "type": "INT",
                  },
                  "period": {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                  "value": "0",
                },
                {
                  "bestValue": true,
                  "leak": "1.0",
                  "metric": {
                    "id": "security_rating",
                    "key": "security_rating",
                    "name": "security_rating",
                    "type": "RATING",
                  },
                  "period": {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                  "value": "0",
                },
              ],
              "name": "index.tsx",
              "path": "src/index.tsx",
              "qualifier": "FIL",
              "value": "0",
            },
            "size": 0,
            "tooltip": <div
              className="text-left"
            >
              <React.Fragment>
                index.tsx
                <br />
              </React.Fragment>
              <React.Fragment>
                ncloc: 0
                <br />
              </React.Fragment>
              <React.Fragment>
                security_remediation_effort: 0
                <br />
              </React.Fragment>
              <React.Fragment>
                vulnerabilities: 0
                <br />
              </React.Fragment>
              <React.Fragment>
                security_rating: \`
              </React.Fragment>
            </div>,
            "x": 0,
            "y": 0,
          },
        ]
      }
      onBubbleClick={[Function]}
      padding={
        [
          25,
          60,
          50,
          60,
        ]
      }
      sizeRange={
        [
          5,
          45,
        ]
      }
      xDomain={
        [
          0,
          100,
        ]
      }
    />
  </div>
  <div
    className="measure-overview-bubble-chart-axis x"
  >
    ncloc
  </div>
  <div
    className="measure-overview-bubble-chart-axis y"
  >
    security_remediation_effort
  </div>
</div>
`;

exports[`should render correctly: default 1`] = `
<div
  className="measure-overview-bubble-chart"
>
  <div
    className="measure-overview-bubble-chart-header"
  >
    <span
      className="measure-overview-bubble-chart-title"
    >
      <div
        className="display-flex-center"
      >
        component_measures.domain_x_overview.Security
        <HelpTooltip
          className="spacer-left"
          overlay={null}
        />
      </div>
    </span>
    <span
      className="measure-overview-bubble-chart-legend"
    >
      <span
        className="note"
      >
        <span
          className="spacer-right"
        >
          component_measures.legend.color_x.security_rating
        </span>
        component_measures.legend.size_x.vulnerabilities
      </span>
      <ColorRatingsLegend
        className="spacer-top"
        filters={{}}
        onRatingClick={[Function]}
      />
    </span>
  </div>
  <div
    className="measure-overview-bubble-chart-content"
  >
    <div
      className="text-center small spacer-top spacer-bottom"
    >
      <ForwardRef(Link)
        to={
          {
            "pathname": "/component_measures",
            "search": "?id=foo&metric=vulnerabilities&view=list",
          }
        }
      >
        component_measures.overview.see_data_as_list
      </ForwardRef(Link)>
    </div>
    <BubbleChart
      displayXGrid={true}
      displayXTicks={true}
      displayYGrid={true}
      displayYTicks={true}
      formatXTick={[Function]}
      formatYTick={[Function]}
      height={500}
      items={
        [
          {
            "color": {
              "fill": "#C6E056",
              "fillTransparent": "rgba(198, 224, 86, 0.2)",
              "stroke": "#809E00",
            },
            "data": {
              "key": "foo:src/index.tsx",
              "leak": "1.0",
              "measures": [
                {
                  "bestValue": true,
                  "leak": "1.0",
                  "metric": {
                    "id": "ncloc",
                    "key": "ncloc",
                    "name": "ncloc",
                    "type": "INT",
                  },
                  "period": {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                  "value": "236",
                },
                {
                  "bestValue": true,
                  "leak": "1.0",
                  "metric": {
                    "id": "security_remediation_effort",
                    "key": "security_remediation_effort",
                    "name": "security_remediation_effort",
                    "type": "DATA",
                  },
                  "period": {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                  "value": "10",
                },
                {
                  "bestValue": true,
                  "leak": "1.0",
                  "metric": {
                    "id": "vulnerabilities",
                    "key": "vulnerabilities",
                    "name": "vulnerabilities",
                    "type": "INT",
                  },
                  "period": {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                  "value": "3",
                },
                {
                  "bestValue": true,
                  "leak": "1.0",
                  "metric": {
                    "id": "security_rating",
                    "key": "security_rating",
                    "name": "security_rating",
                    "type": "RATING",
                  },
                  "period": {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                  "value": "2",
                },
              ],
              "name": "index.tsx",
              "path": "src/index.tsx",
              "qualifier": "FIL",
              "value": "3",
            },
            "size": 3,
            "tooltip": <div
              className="text-left"
            >
              <React.Fragment>
                index.tsx
                <br />
              </React.Fragment>
              <React.Fragment>
                ncloc: 236
                <br />
              </React.Fragment>
              <React.Fragment>
                security_remediation_effort: 10
                <br />
              </React.Fragment>
              <React.Fragment>
                vulnerabilities: 3
                <br />
              </React.Fragment>
              <React.Fragment>
                security_rating: B
              </React.Fragment>
            </div>,
            "x": 236,
            "y": 10,
          },
        ]
      }
      onBubbleClick={[Function]}
      padding={
        [
          25,
          60,
          50,
          60,
        ]
      }
      sizeRange={
        [
          5,
          45,
        ]
      }
    />
  </div>
  <div
    className="measure-overview-bubble-chart-axis x"
  >
    ncloc
  </div>
  <div
    className="measure-overview-bubble-chart-axis y"
  >
    security_remediation_effort
  </div>
</div>
`;

exports[`should render correctly: only showing first 500 files 1`] = `
<div
  className="measure-overview-bubble-chart"
>
  <div
    className="measure-overview-bubble-chart-header"
  >
    <span
      className="measure-overview-bubble-chart-title"
    >
      <div
        className="display-flex-center"
      >
        component_measures.domain_x_overview.Security
        <HelpTooltip
          className="spacer-left"
          overlay={null}
        />
      </div>
      <div
        className="note spacer-top"
      >
        (
        component_measures.legend.only_first_500_files
        )
      </div>
    </span>
    <span
      className="measure-overview-bubble-chart-legend"
    >
      <span
        className="note"
      >
        <span
          className="spacer-right"
        >
          component_measures.legend.color_x.security_rating
        </span>
        component_measures.legend.size_x.vulnerabilities
      </span>
      <ColorRatingsLegend
        className="spacer-top"
        filters={{}}
        onRatingClick={[Function]}
      />
    </span>
  </div>
  <div
    className="measure-overview-bubble-chart-content"
  >
    <div
      className="text-center small spacer-top spacer-bottom"
    >
      <ForwardRef(Link)
        to={
          {
            "pathname": "/component_measures",
            "search": "?id=foo&metric=vulnerabilities&view=list",
          }
        }
      >
        component_measures.overview.see_data_as_list
      </ForwardRef(Link)>
    </div>
    <BubbleChart
      displayXGrid={true}
      displayXTicks={true}
      displayYGrid={true}
      displayYTicks={true}
      formatXTick={[Function]}
      formatYTick={[Function]}
      height={500}
      items={
        [
          {
            "color": {
              "fill": "#C6E056",
              "fillTransparent": "rgba(198, 224, 86, 0.2)",
              "stroke": "#809E00",
            },
            "data": {
              "key": "foo:src/index.tsx",
              "leak": "1.0",
              "measures": [
                {
                  "bestValue": true,
                  "leak": "1.0",
                  "metric": {
                    "id": "ncloc",
                    "key": "ncloc",
                    "name": "ncloc",
                    "type": "INT",
                  },
                  "period": {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                  "value": "236",
                },
                {
                  "bestValue": true,
                  "leak": "1.0",
                  "metric": {
                    "id": "security_remediation_effort",
                    "key": "security_remediation_effort",
                    "name": "security_remediation_effort",
                    "type": "DATA",
                  },
                  "period": {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                  "value": "10",
                },
                {
                  "bestValue": true,
                  "leak": "1.0",
                  "metric": {
                    "id": "vulnerabilities",
                    "key": "vulnerabilities",
                    "name": "vulnerabilities",
                    "type": "INT",
                  },
                  "period": {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                  "value": "3",
                },
                {
                  "bestValue": true,
                  "leak": "1.0",
                  "metric": {
                    "id": "security_rating",
                    "key": "security_rating",
                    "name": "security_rating",
                    "type": "RATING",
                  },
                  "period": {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                  "value": "2",
                },
              ],
              "name": "index.tsx",
              "path": "src/index.tsx",
              "qualifier": "FIL",
              "value": "3",
            },
            "size": 3,
            "tooltip": <div
              className="text-left"
            >
              <React.Fragment>
                index.tsx
                <br />
              </React.Fragment>
              <React.Fragment>
                ncloc: 236
                <br />
              </React.Fragment>
              <React.Fragment>
                security_remediation_effort: 10
                <br />
              </React.Fragment>
              <React.Fragment>
                vulnerabilities: 3
                <br />
              </React.Fragment>
              <React.Fragment>
                security_rating: B
              </React.Fragment>
            </div>,
            "x": 236,
            "y": 10,
          },
        ]
      }
      onBubbleClick={[Function]}
      padding={
        [
          25,
          60,
          50,
          60,
        ]
      }
      sizeRange={
        [
          5,
          45,
        ]
      }
    />
  </div>
  <div
    className="measure-overview-bubble-chart-axis x"
  >
    ncloc
  </div>
  <div
    className="measure-overview-bubble-chart-axis y"
  >
    security_remediation_effort
  </div>
</div>
`;
